File: app/src/main/kotlin/me/tomasan7/jecnadesktop/repository/WebGradesRepository.java
package me.tomasan7.jecnadesktop.repository;

import me.tomasan7.jecnadesktop.data.Grades;
import me.tomasan7.jecnadesktop.parser.parsers.HTMLGradesParser;
import me.tomasan7.jecnadesktop.web.JecnaWebClient;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;

/**
 * Retrieves {@link Grades} from the Ječná web.
 */
public class WebGradesRepository implements GradesRepository
{
	private static final String WEB_PATH = "/score/student";

	private final JecnaWebClient webClient;
	private final HTMLGradesParser gradesParser = new HTMLGradesParser();

	public WebGradesRepository (JecnaWebClient webClient)
	{
		this.webClient = webClient;
	}

	@Override
	public Grades queryGrades ()
	{
		try
		{
			return gradesParser.parse(webClient.queryStringBody(WEB_PATH).get());
		}
		catch (ExecutionException | InterruptedException e)
		{
			throw new RuntimeException(e);
		}
	}

	@Override
	public CompletableFuture<Grades> queryGradesAsync ()
	{
		return webClient.queryStringBody(WEB_PATH)
						.thenApply(gradesParser::parse);
	}
}

File: src/main/kotlin/me/tomasan7/jecnadesktop/repository/WebGradesRepository.kt
package me.tomasan7.jecnadesktop.repository

import io.ktor.http.*
import me.tomasan7.jecnadesktop.data.Grades
import me.tomasan7.jecnadesktop.data.SchoolYear
import me.tomasan7.jecnadesktop.parser.parsers.HtmlGradesParser
import me.tomasan7.jecnadesktop.util.JecnaPeriodEncoder
import me.tomasan7.jecnadesktop.util.JecnaPeriodEncoder.jecnaEncode
import me.tomasan7.jecnadesktop.web.JecnaWebClient
import me.tomasan7.jecnadesktop.web.append

/**
 * Retrieves [Grades] from the Ječná web.
 */
class WebGradesRepository(private val webClient: JecnaWebClient) : GradesRepository
{
    private val gradesParser = HtmlGradesParser()

    override suspend fun queryGrades() = gradesParser.parse(webClient.queryStringBody(WEB_PATH))

    override suspend fun queryGrades(schoolYear: SchoolYear, firstHalf: Boolean) = gradesParser.parse(webClient.queryStringBody(WEB_PATH, Parameters.build {
        append(schoolYear.jecnaEncode())
        append(JecnaPeriodEncoder.encodeSchoolYearHalf(firstHalf))
    }))

    companion object
    {
        private const val WEB_PATH = "/score/student"
    }
}